Refactor Therapy Settings Editors

marv-out opened the pull request:
# Refactor Therapy Settings Editors to Use Shared Component

## Summary
This PR refactors all four therapy settings editors (Glucose Targets, Basal Rates, Carb Ratios, and Insulin Sensitivities) to use the unified `TherapySettingEditorView` component, bringing consistency with the onboarding flow.

## Changes Made

1. Fixed some problems in onboarding where the user can't really add multiple entries and edit them
2. Refactored All Therapy Settings Editors to match the onboarding UI using the `TherapySettingEditorView` component

marv-out committed:
Merge branch 'dev' of github.com:marv-out/Trio into dev

marv-out committed:
refine therapy settings UI

marv-out committed:
scroll to bottom when adding a new item

marv-out committed:
refine therapy settings UI in app

marv-out committed:
refine therapy settings UI in app (carb ratio)

marv-out committed:
refine therapy settings UI in app (carb ratio)

marv-out committed:
refine therapy settings UI in app (isf)

marv-out committed:
add descriptions from onboarding view

marv-out committed:
tint color red for swipe to delete action

marv-out committed:
Merge branch 'dev' into therapy

marv-out committed:
add descriptions from onboarding view

marv-out committed:
Merge branch 'therapy' of github.com:marv-out/Trio into therapy

Sjoerd-Bo3 commented:
Already great rework, I think it's much clearer already!

Few things I encountered while testing in settings:
- It does allow you to tap the 'add' button even when there is a 23:30 entry, which results in not an extra row being added.

<video width="280" height="800" src="https://github.com/user-attachments/assets/29e5991c-4dd3-40ca-947e-aa7aaa6ea8b0.mp4" controls preload></video>

- When changing an entry, and sliding one of the values to another entry. It auto closes pretty abruptly. Also I have no chance to change the other item. Is this intended? I think it would be more logical to have it open until you close it by clicking the row?

<video width="280" height="800" src="https://github.com/user-attachments/assets/10c3ba50-16a4-4ed1-a780-dc0679139078.mp4" controls preload></video>

dnzxy commented:
Thanks for taking on the port of my Onboarding work and refactoring the Therapy Settings, plus fixing iOS 26 SDK bugs.

As for the UI: this is the expected look of the chart and content (Screenshot taken from Onboarding).

<img width="283" height="236" alt="image" src="https://github.com/user-attachments/assets/508c2322-395d-4341-9425-70cfd772a5e1" />

- The text and double headlines are unnecessary.
- Users have already gone through Onboarding, they know their settings and what each setting does. 

There is a change to `Info.plist` that seems unnecessary.

MikePlante1 commented:
> There is a change to `Info.plist` that seems unnecessary.

I think that’s one of those things that if you build with Xcode 26 then it makes that change when you build. But then if you build with Xcode 16.4 it reverts that change.

marv-out commented:
>  The text and double headlines are unnecessary.

I just wanted to recreate the onboarding UI in the first step and get some feedback on it. The double headlines aren’t really necessary, but I wanted to include that emoji in the navigation title — which isn’t quite possible without a custom nav title. As you know, I’m a fan of those emojis (even though I know you’re not xD).

Anyway, which texts should stay, and which ones can go? I actually like the explanatory texts in the CR Editor view and the ISF view.

> There is a change to Info.plist that seems unnecessary.

As Mike said I haven't changed anything there. Must have been Xcode 26

marv-out commented:
> When changing an entry, and sliding one of the values to another entry. It auto closes pretty abruptly. Also I have no chance to change the other item. Is this intended? I think it would be more logical to have it open until you close it by clicking the row?

No, thats not intended. I will have a look

> It does allow you to tap the 'add' button even when there is a 23:30 entry, which results in not an extra row being added.

Ok, we can disable the button in that case

dnzxy commented:
> The double headlines aren’t really necessary, but I wanted to include that emoji in the navigation title

I’d suggest to keep it default SwiftUI, and remove the double title, so the one with the icon.

> which texts should stay, and which ones can go? 

Imho all of them. The goal would be to replace the current setup (Add button top right corner, chart, then rates) where you need to tap on a table item, it opens a NavView, you need to change time and rate etc. and make it into the inline-editing UI we have in Onboarding. 

Whether we keep the Add button above the list (and sticky) have it be a toolbar item could be discussed.

> As Mike said I haven't changed anything there. Must have been Xcode 26

Thanks for clarifying that to you and Mike. My bad. 

The current (before PR) therapy setting views also have a small warning when the schedule becomes "full". This may be something to keep, possibly. Open for discussion. 

<img width="333" alt="image" src="https://github.com/user-attachments/assets/7a4d8adf-1c1c-4fdb-8a9a-230f173fc641" />

The main thing we need to fix and we fixed that with what was built for Onboarding, is the whole back and forth navigation stuff. That a) bad UX and b) buggy for iOS26, as it doesn’t keep changed entry values.

Sjoerd-Bo3 commented:
I like the message about full rates.

marv-out commented:
> Imho all of them. 

just to clarify, you are talking about removing (texts) here, aren't you?

> Whether we keep the Add button above the list (and sticky) have it be a toolbar item could be discussed.

when we change it here we also need to change it in onboarding. I like it as it is right now, but I needed to change it with this PR to move it outside the list

> The current (before PR) therapy setting views also have a small warning when the schedule becomes "full".

I like that, will add it back

dnzxy commented:
> > Imho all of them.
> 
> just to clarify, you are talking about removing (texts) here, aren't you?
> 
Yes, thats why I quoted you.

> > Whether we keep the Add button above the list (and sticky) have it be a toolbar item could be discussed.
> 
> when we change it here we also need to change it in onboarding. 

Not really, as the UI in Onboarding is different and you cannot fit a toolbar button there. I was specifically talking about this PR and therapy settings.

marv-out commented:
> Yes, thats why I quoted you.

Yeah, but I asked which texts could be deleted AND also which texts should stay and you simply said `all` 😂 nvm

> Not really, as the UI in Onboarding is different and you cannot fit a toolbar button there. I was specifically talking about this PR and therapy settings.

Well...doesn't matter that much to me where this button is. The most important is that its no longer in the List and doesn't scroll away when there are too many items in the List. One argument in favor of the toolbar would be that its more Swift-like

dnzxy commented:
> Yeah, but I asked which texts could be deleted AND also which texts should stay and you simply said `all` 😂 nvm

My suggestion was to: Remove all texts. Remove one of the duplicate titles.

> One argument in favor of the toolbar would be that its more Swift-like

Agree absolutely, but we don't have that "screen real estate" in Onboarding. Not really relevant for this PR / your work here though.

dnzxy commented:
Side note: There are merge conflicts as latest translations were merged to dev today.

MikePlante1 committed:
Merge branch 'dev' into pr/819

MikePlante1 commented:
> Side note: There are merge conflicts as latest translations were merged to dev today.

@marv-out Are you okay with me doing and pushing this merge to make sure some of the onboarding localization changes don't get reverted?

MikePlante1 committed:
Revert the revert of #813 and apply Xcode 16.4

 fix-its after merging upstream

MikePlante1 commented:
My language was imprecise. To clarify — I was tested merging 0.6.0.5 in with Xcode 16.4 vs Xcode 26 because development should be done in Xcode 16.4 given that's what Trio's fastlane is currently using... and in doing so I discovered most of #813 got reverted when it was merged earlier in this PR. So I'd like to correct that:

https://github.com/marv-out/Trio/compare/therapy...MikePlante1:819?expand=1

marv-out commented:
> Are you okay with me doing and pushing this merge to make sure some of the onboarding localization changes don't get reverted?

Yes xD

dnzxy commented:
Have the two "issues" in [this comment](https://github.com/nightscout/Trio/pull/819#issuecomment-3423141455) been addressed yet? I'm not sure, I don't see any commits. @marv-out ? 👀

dnzxy committed:
Merge branch 'dev' into therapy

marv-out committed:
disable add button when last entry is at 11.30 p.m.

marv-out committed:
remove double headers and texts

marv-out committed:
readd basal coverage text

marv-out committed:
fix picker being closed abruptly when changing a value

This was due to the generation of new UUIDs which made selectedItemID invalid and led to the picker being closed

marv-out committed:
Merge branch 'dev' of github.com:nightscout/Trio into therapy

marv-out commented:
Ok please retest/ review again @Sjoerd-Bo3 @dnzxy @MikePlante1 

I've removed the texts, re-added the basal coverage box, fixed the picker behaviour and disabled the add button when the last entries' time was at 11:30 p.m.

should be good to go imo (unless you wanna move the add button to the toolbar in trailing position, haven't done that yet)

MikePlante1 commented:
Not sure why part of #813 got reverted again... but I'll push a fix for that shortly, along with some Xcode 16.4 fix-its, including the linting that was introduced with #705 (0.6.0.7).

Missing space between value and units and ISF's units should be mg/dL/U:

<img width="333" src="https://github.com/user-attachments/assets/1148fc7b-03af-4c1d-9c6d-1620cf87f5b9" />

EDIT: I figured it out. I missed this part in my previous revert-revert-813 commit. Oops.

MikePlante1 committed:
Revert the revert of #813

plus apply linting introduced from #705 (0.6.0.7)

marv-out committed:
Merge branch 'dev' of github.com:nightscout/Trio into therapy

marv-out commented:
Whats the status of this PR? Is it good to go?

dnzxy commented:
Will try and get a thorough review in tonight (pacific time, so in roughly 8-10 hrs).

dnzxy committed:
Cleanup + Fix UI Styling for button row

dnzxy commented:
I've cleaned up the UI a bit and pushed a few adjustments. 
Looks good to me so far :)

MikePlante1 commented on Trio/Sources/Modules/ISFEditor/View/ISFEditorRootView.swift:
```diff
@@ -67,143 +69,100 @@ extension ISFEditor {
-                                            .timeValues[i]
-                                    ))
-                            ).tag(i)
+                                    Text("Swipe to delete a single entry. Tap on it, to edit its time or rate.")
```
> `time or rate.`

*`time or value.`

dnzxy replied:
Addressed. 

I actually cross compared and in the current implementation (as of latest `dev`), it says `rate` for all the settings, so targets, basal, carb ratio and isf. It now says value everywhere. It only seemed to say `rate` for ISF, as I had already adjusted it to say `value` everywhere else.

MikePlante1 commented:
I find it strange how Basal Rates all get set to 0.05 U/hr when a pump is not connected. Maybe this isn't something we can easily fix, though, since unlike other therapy settings, editing basal rates requires a pump to be connected.

---

<img width="425" height="126" alt="Screenshot 2025-11-14 at 06 02 03" src="https://github.com/user-attachments/assets/fe3e3e0a-a6ec-4075-8cf7-b0399c283c1c" />

This appears in Therapy settings in the app, but not in Onboarding.

---

<img width="1377" height="1259" alt="Screenshot 2025-11-14 at 05 18 20" src="https://github.com/user-attachments/assets/c665f859-a217-4822-8b51-48d642bc1eef" />

I went through onboarding in mg/dL, set the target to 180 mg/dL, went back switched to mmol/L, and Trio crashes out.

I reopened the app, which reset onboarding, started with mmol/L this time, set a target of 10.0 mmol/L, went back to switch it to mg/dL, then went forward in onboarding and it used 132 mg/dL instead of 180 mg/dL. Also tried 8.0 mmol/L, which it converted to 112 mg/dL instead of 144 mg/dL.

Same thing with ISF too.

This wasn't introduced with this PR, though, as I checked `dev` 0.6.0.9 and got the same results.

---

Another thing not introduced in this PR, but I find it odd that setting a Scheduled Basal Rate > 2 U/hr doesn't either automatically increase `Maximum Basal Rate` to the highest rate or just give an error and force the user to edit it themselves.

---

Will resuming testing after I board my flight...

dnzxy commented:
> I find it strange how Basal Rates all get set to 0.05 U/hr when a pump is not connected. Maybe this isn't something we can easily fix, though, since unlike other therapy settings, editing basal rates requires a pump to be connected.

Good find, good testing haha. That’s why for the Onboarding we have introduced the pump selection step. The pump you choose there is just a mock pump that is used to get an increment value for the subsequent settings screens. We don’t really have a good way of doing this for basal once onboarded when the user has no onboarded (i.e. instantiated PumpManager via one of the submodules) pump. It’s annoying but not sure if it’s important. 

We _could_ persist the pump selection the user chose during Onboarding but what happens if the user then uses a pod and then switches their pump. 

When the user has a pump onboarded, we set the increment correctly. It’s a bit of a shitty situation, tbh. 

I‘ll fix the typo and would open new issues for the other stuff, plus wait for whatever you may find later :)

MikePlante1 commented:
Nothing else came up in my testing, so once that typo is fixed, I think this is good to merge.

dnzxy committed:
feat(i18n): Fix typo in hint #hackdiabetes2025
